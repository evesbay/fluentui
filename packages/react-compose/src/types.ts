import * as React from 'react';

// tslint:disable-next-line:interface-name
export interface ShorthandConfig<P> {
  mappedProp?: keyof P;
  mappedArrayProp?: keyof P;
  allowsJSX?: boolean;
}

//
// "as" type safety
//

export type PropsOfElement<
  // tslint:disable-next-line:no-any
  E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
> = JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;

// tslint:disable-next-line:interface-name
export interface ComponentWithAs<E extends React.ElementType = 'div', P = {}> extends React.FunctionComponent {
  <EE extends React.ElementType = E>(
    props: Omit<PropsOfElement<EE>, 'as' | keyof P> & { as?: EE } & P,
  ): JSX.Element | null;
  displayName?: string;

  defaultProps?: Partial<P & { as: E }>;
  propTypes?: React.WeakValidationMap<P> & {
    // tslint:disable-next-line:no-any
    as: React.Requireable<string | ((props: any, context?: any) => any) | (new (props: any, context?: any) => any)>;
  };
}

//
// Compose types
//

export type ComposedComponent<P = {}> = React.FunctionComponent<P> & {
  fluentComposeConfig: ComposePreparedOptions;
};

export type InputComposeComponent<P = {}> = React.FunctionComponent<P> & {
  fluentComposeConfig?: ComposePreparedOptions;
};

export type Input<T extends React.ElementType = 'div', P = {}> =
  | InputComposeComponent<P>
  | ComposeRenderFunction<T, P & { as?: React.ElementType }>;

export type ComposeRenderFunction<T extends React.ElementType = 'div', P = {}> = (
  props: P,
  ref: React.Ref<T extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[T] : T>,
  composeOptions: ComposePreparedOptions,
) => React.ReactElement | null;

export type ComposeOptions<InputProps = {}, InputStylesProps = {}, ParentStylesProps = {}> = {
  className?: string;
  classes?: ClassDictionary;
  displayName?: string;

  mapPropsToStylesProps?: (props: ParentStylesProps & InputProps) => InputStylesProps;

  handledProps?: (keyof InputProps | 'as')[];
  overrideStyles?: boolean;

  slots?: Record<string, React.ElementType>;

  mapPropsToSlotProps?: (props: InputProps) => Record<string, object>;

  shorthandConfig?: ShorthandConfig<InputProps>;
};

export type ClassDictionary = {
  [key: string]: string;
};

export type ComposePreparedOptions<Props = {}> = {
  className: string;
  classes: ClassDictionary;
  displayName: string;
  displayNames: string[];

  mapPropsToStylesPropsChain: ((props: object) => object)[];
  render: ComposeRenderFunction;

  handledProps: (keyof Props)[];
  overrideStyles: boolean;

  slots: Record<string, React.ElementType> & { __self: React.ElementType };
  mapPropsToSlotPropsChain: ((props: Props) => Record<string, object>)[];

  resolveSlotProps: <P>(props: P) => Record<string, object>;
  shorthandConfig: ShorthandConfig<Props>;
};
