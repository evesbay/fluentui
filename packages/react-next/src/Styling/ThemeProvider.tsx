import * as React from 'react';
import { ThemeProvider as ReactThemeProvider, ThemeProviderProps } from '@fluentui/react-theme-provider';
import { useCustomizationSettings } from '../Utilities';
import { convertLegacyTheme } from './convertLegacyTheme';
import { ITheme } from '../Styling';

export { ThemeProviderProps } from '@fluentui/react-theme-provider';

/**
 * A wrapper of ThemeProvider from react-theme-provider package.
 *
 * It also updates the theme when Customizations changes, which ensures backward compatibility with legacy ways of
 * providing theme (e.g. loadTheme, Customizations.applySettings).
 */
export const ThemeProvider: React.FunctionComponent<ThemeProviderProps & {
  ref?: React.Ref<HTMLDivElement>;
}> = props => {
  const legacyTheme = useThemeCustomizationSettings();
  const theme = props.theme || (legacyTheme && convertLegacyTheme(legacyTheme));

  return <ReactThemeProvider {...props} theme={theme} />;
};

function useThemeCustomizationSettings(): ITheme {
  return useCustomizationSettings(['theme']).theme;
}
