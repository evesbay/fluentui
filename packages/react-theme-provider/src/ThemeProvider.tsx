import * as React from 'react';
import cx from 'classnames';
import { useStylesheet } from '@fluentui/react-stylesheets';
import { tokensToStyleObject } from './tokensToStyleObject';
import { ThemeContext } from './ThemeContext';
import { Theme, ThemePrepared } from './types';
import { mergeThemes } from './mergeThemes';
import { useTheme } from './useTheme';
import * as classes from './ThemeProvider.scss';

/**
 * Props for the ThemeProvider component.
 */
export interface ThemeProviderProps extends React.HTMLAttributes<HTMLDivElement> {
  /**
   * Defines the theme provided by the user.
   */
  theme?: Theme;
}

/**
 * ThemeProvider, used for providing css variables and registering stylesheets.
 */
export const ThemeProvider = React.forwardRef<HTMLDivElement, ThemeProviderProps>(
  (
    { theme, style, className, ...rest }: React.PropsWithChildren<ThemeProviderProps>,
    ref: React.Ref<HTMLDivElement>,
  ) => {
    // Pull contextual theme.
    const parentTheme = useTheme();

    // Merge the theme only when parent theme or props theme mutates.
    const fullTheme = React.useMemo<ThemePrepared>(() => mergeThemes(parentTheme, theme), [parentTheme, theme]);

    // Generate the inline style object only when merged theme mutates.
    const inlineStyle = React.useMemo<React.CSSProperties>(
      () => tokensToStyleObject(fullTheme.tokens, undefined, { ...style }),
      [fullTheme, style],
    );

    const rootClass = cx(className, classes.root) || undefined;

    // Register stylesheets as needed.
    useStylesheet(fullTheme.stylesheets);

    // Provide the theme in case it's required through context.
    // tslint:disable:jsx-ban-props
    return (
      <ThemeContext.Provider value={fullTheme}>
        <div {...rest} ref={ref} className={rootClass} style={inlineStyle} />
      </ThemeContext.Provider>
    );
    // tslint:enable:jsx-ban-props
  },
);

ThemeProvider.displayName = 'ThemeProvider';
