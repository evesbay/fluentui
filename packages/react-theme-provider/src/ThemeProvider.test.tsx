import * as React from 'react';
import { ThemeProvider } from './ThemeProvider';
import * as renderer from 'react-test-renderer';
import { Theme } from './types';
import { useTheme } from './useTheme';
import { mount } from 'enzyme';
import { mergeThemes } from './mergeThemes';

const lightTheme: Theme = mergeThemes({
  stylesheets: [],
  tokens: {
    body: {
      fill: 'white',
      text: 'black',
    },
  },
});

const darkTheme: Theme = mergeThemes({
  tokens: {
    body: {
      fill: 'black',
      text: 'white',
    },
  },
});

describe('ThemeProvider', () => {
  it('renders a div', () => {
    const component = renderer.create(<ThemeProvider>Hello</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a div with styling', () => {
    const component = renderer.create(<ThemeProvider theme={lightTheme}>Hello</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nested themes', () => {
    const component = renderer.create(
      <ThemeProvider theme={lightTheme}>
        <div>Light theme</div>
        <ThemeProvider theme={darkTheme}>
          <div>Dark theme</div>
        </ThemeProvider>
      </ThemeProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('provide the theme through context', () => {
    let resolvedTheme: Theme | undefined = undefined;
    const TestComponent = () => {
      resolvedTheme = useTheme();

      return null;
    };

    mount(
      <ThemeProvider theme={lightTheme}>
        <TestComponent />
      </ThemeProvider>,
    );

    expect(resolvedTheme).toEqual(lightTheme);
  });
});
