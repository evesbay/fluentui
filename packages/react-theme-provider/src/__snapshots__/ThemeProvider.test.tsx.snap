// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders a div 1`] = `
<div
  style={
    Object {
      "--body-divider": "#eee",
      "--body-fill": "white",
      "--body-link": "blue",
      "--body-subText": "#333",
      "--body-text": "black",
    }
  }
>
  Hello
</div>
`;

exports[`ThemeProvider renders a div with styling 1`] = `
<div
  style={
    Object {
      "--body-divider": "#eee",
      "--body-fill": "white",
      "--body-link": "blue",
      "--body-subText": "#333",
      "--body-text": "black",
    }
  }
>
  Hello
</div>
`;

exports[`ThemeProvider renders nested themes 1`] = `
<div
  style={
    Object {
      "--body-divider": "#eee",
      "--body-fill": "white",
      "--body-link": "blue",
      "--body-subText": "#333",
      "--body-text": "black",
    }
  }
>
  <div>
    Light theme
  </div>
  <div
    style={
      Object {
        "--body-divider": "#eee",
        "--body-fill": "black",
        "--body-link": "blue",
        "--body-subText": "#333",
        "--body-text": "white",
      }
    }
  >
    <div>
      Dark theme
    </div>
  </div>
</div>
`;
